contracts/token/ERC20/extensions/ERC7540.sol

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {IERC7540} from "../../../interfaces/IERC7540.sol";
+import {IERC20} from "../../../interfaces/IERC20.sol";
+import {ERC4626} from "./ERC4626.sol";
+import {SafeERC20} from "../utils/SafeERC20.sol";
+import {Math} from "../../../utils/math/Math.sol";
+import {ReentrancyGuard} from "../../../utils/ReentrancyGuard.sol";
+
+/**
+ * @dev Abstract implementation of the ERC-7540 standard, extending ERC-4626.
+ */
+abstract contract ERC7540 is ERC4626, IERC7540, ReentrancyGuard {
+    using SafeERC20 for IERC20;
+    using Math for uint256;
+
+    struct Request {
+        uint256 amount;
+        uint256 claimable;
+    }
+
+    // Mappings to track pending and claimable requests
+    mapping(address => mapping(uint256 => Request)) internal _pendingDepositRequests;
+    mapping(address => mapping(uint256 => Request)) internal _pendingRedeemRequests;
+
+    mapping(address => mapping(address => bool)) private _operators;
+
+    /**
+     * @dev Set the underlying asset contract.
+     */
+    constructor(IERC20 asset) ERC4626(asset) {}
+
+    /**
+     * @dev Creates a new deposit request.
+     */
+    function requestDeposit(
+        uint256 assets,
+        address controller,
+        address owner
+    ) external override nonReentrant returns (uint256 requestId) {
+        require(assets > 0, "ERC7540: assets must be greater than zero");
+        require(owner == msg.sender || isOperator(owner, msg.sender), "ERC7540: unauthorized");
```

Comment:
Two things here:

- we avoid using `msg.sender` and instead use `_msgSender()` provided by `Context`. This is necessary for supporting ERC-2771meta-tx.
- We avoid using revert strings. Instead, we only use custom errors that can give more context. Please see [ERC-6093](https://eips.ethereum.org/EIPS/eip-6093)

Reply:
Ok i didnt know that, tnx for clarifying it. Its fixed @Amxx